Update Great Expectations Guide to use V3 API Operator. (#321)
astronomer/airflow-guides: guides/airflow-great-expectations.md

1. Initialized a Great Expectations project
2. Configured at least one Datasource `my_datasource`
3. Created at least one Expectation Suite `my_suite`
4. Created a Checkpoint `my_checkpoint`

If you set up a project manually, you will see a `great_expectations` directory which contains several sub-directories, as well as the `great_expectations.yml` configuration file. If you cloned the demo repository, the `great_expectations` directory can be found under `include/`.

> Note: The `GreatExpectationsOperator` requires Airflow >=2.1, and you will need to change the value of `enable_xcom_pickling` to `true` in your airflow.cfg. If you are using an Astronomer project structure, add `ENV AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True` to your Dockerfile. If you are working from the demo repository, this step has already been completed for you.

## Use Case: Great Expectations Operator

Now that we've set up our system to work with Great Expectations, we can start exploring how to use it in our DAGs. The current Great Expectations provider version uses the [Great Expectations V3 API](https://docs.greatexpectations.io/docs/).

### Configuration

> The `GreatExpectationsOperator` runs a Checkpoint against a given dataset. Before you start writing your DAG, make sure you have a Data Context and Checkpoint configured. These can be defined as dictionaries containing the necessary [Data Context](https://github.com/astronomer/airflow-data-quality-demo/blob/main/include/great_expectations/configs/snowflake_configs.py#L14) and [Checkpoint](https://github.com/astronomer/airflow-data-quality-demo/blob/main/include/great_expectations/configs/snowflake_configs.py#L99) fields and then imported into the DAG.

Reviewer: ```suggestion
To validate your data, the `GreatExpectationsOperator` runs a Checkpoint against your dataset. Before you start writing your DAG, make sure you have a Data Context and Checkpoint configured. To do this, define dictionaries containing the necessary [Data Context](https://github.com/astronomer/airflow-data-quality-demo/blob/main/include/great_expectations/configs/snowflake_configs.py#L14) and [Checkpoint](https://github.com/astronomer/airflow-data-quality-demo/blob/main/include/great_expectations/configs/snowflake_configs.py#L99) fields and import those dictionaries into your DAG.
```

[Checkpoints](https://docs.greatexpectations.io/docs/reference/checkpoints_and_actions) provide a convenient abstraction for bundling the validation of a Batch (or Batches) of data against an Expectation Suite (or several), as well as the actions that should be taken after the validation.

Our [demo repository](https://github.com/astronomer/airflow-data-quality-demo/) uses the following configuration:

- The `great_expectations` directory is accessible by your DAG, as it is loaded into Docker as part of the `include` directory.
- The Great Expectations provider is installed when you run `astro dev start`, as it is part of `requirements.txt`. For Astronomer projects, a provider version >=`0.1.1` is required. If you are not using an Astronomer environment, install the Great Expectations provider in your environment manually:

    ```bash
    pip install great_expectations airflow-provider-great-expectations>=0.1.0
    ```

    It’s recommended to specify a version when installing the package. To make use of the latest Great Expectations V3 API, you need to specify a version >= `0.1.0`.

### Using the Great Expectations Operator

> 1. Import the operator in your DAG file. You may also need to import the `DataContextConfig`, `CheckpointConfig`, or `BatchRequest` classes as well, depending on how you're using the operator. To import the Great Expectations provider and config and batch classes in a given DAG, add the following line to the top of the DAG file in your `dags` directory:

Reviewer: ```suggestion
1. Import the operator in your DAG file. You might also need to import the `DataContextConfig`, `CheckpointConfig`, or `BatchRequest` classes depending on how you're using the operator. To import the Great Expectations provider, configurations, and batch classes in a given DAG, add the following line to the top of the DAG file in your `dags` directory:
```

For more information about possible parameters and examples, see the [README in the provider repository](https://github.com/great-expectations/airflow-provider-great-expectations) and the [example DAG in the provider package](https://registry.astronomer.io/dags/example-great-expectations-dag).

### Connections and Backends

The `GreatExpectationsOperator` can run a checkpoint on a dataset stored in any backend compatible with Great Expectations, e.g. BigQuery, MSSQL, MySQL, PostgreSQL, Redshift, Snowflake, SQLite, Athena. All that’s needed to get the Operator to point at an external dataset is to set up an [Airflow Connection](https://www.astronomer.io/guides/connections) to the `datasource`, and add the connection to your Great Expectations project, e.g. [using the CLI to add a Postgres backend](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/database/postgres). Then, if using a `DataContextConfig` or `CheckpointConfig`, ensure that the `"datasources"` field refers to your backend connection name.

## Conclusion

> In this guide, we presented a brief overview of Great Expectations and how to use the provider operator to create Great Expectations Airflow Tasks. For more examples on how to use the `GreatExpectationsOperator` as part of an ELT pipeline, check out the [Snowflake](https://registry.astronomer.io/dags/simple-great-expectations-snowflake-el), [BigQuery](https://registry.astronomer.io/dags/simple-great-expectations-bigquery-el/), and [Redshift](https://registry.astronomer.io/dags/simple-great-expectations-redshift-el) examples on the [Astronomer Registry](https://registry.astronomer.io/).

Reviewer: ```suggestion
In this guide, we presented a brief overview of Great Expectations and explained how to use the provider operator to create Great Expectations Airflow tasks. For more examples on how to use the `GreatExpectationsOperator` as part of an ELT pipeline, check out the [Snowflake](https://registry.astronomer.io/dags/simple-great-expectations-snowflake-el), [BigQuery](https://registry.astronomer.io/dags/simple-great-expectations-bigquery-el/), and [Redshift](https://registry.astronomer.io/dags/simple-great-expectations-redshift-el) examples on the [Astronomer Registry](https://registry.astronomer.io/).
```

If you wish to use your own Great Expectations project along with this guide, ensure you have completed the following steps:

1. Initialized a Great Expectations project
2. Configured at least one Datasource `my_datasource`
3. Created at least one Expectation Suite `my_suite`
4. Created a Checkpoint `my_checkpoint`

If you set up a project manually, you will see a `great_expectations` directory which contains several sub-directories, as well as the `great_expectations.yml` configuration file. If you cloned the demo repository, the `great_expectations` directory can be found under `include/`.

> Note: The `GreatExpectationsOperator` requires Airflow >=2.1, and you will need to change the value of `enable_xcom_pickling` to `true` in your airflow.cfg. If you are using an Astronomer project structure, add `ENV AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True` to your Dockerfile. If you are working from the demo repository, this step has already been completed for you.

## Use Case: Great Expectations Operator

> Now that we've set up our system to work with Great Expectations, we can start exploring how to use it in our DAGs. The current Great Expectations provider version uses the Great Expectations V3 API.

Author: Yes - good catch, I forgot the link. Thanks!

### Configuration

The `GreatExpectationsOperator` runs a Checkpoint against a given dataset. Before you start writing your DAG, make sure you have a Data Context and Checkpoint configured.

A [Data Context](https://docs.greatexpectations.io/docs/reference/data_context) represents a Great Expectations project. It organizes storage and access for Expectation Suites, data sources, notification settings, and data fixtures.

[Checkpoints](https://docs.greatexpectations.io/docs/reference/checkpoints_and_actions) provide a convenient abstraction for bundling the validation of a Batch (or Batches) of data against an Expectation Suite (or several), as well as the actions that should be taken after the validation.

> Our [demo repository](https://github.com/astronomer/airflow-data-quality-demo/) uses the following configuration:

Author: Added a blurb in the beginning to link out to code snippets as they're a bit lengthy

    from great_expectations.data_context.types.base import (
        DataContextConfig,
        CheckpointConfig
    )
    ```

2. Create a task using the [`GreatExpectationsOperator`](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator).

To use the operator in the DAG, define an instance of the `GreatExpectationsOperator` class and assign it to a variable. In the following example, we define two different instances of the operator to complete two different steps in a data quality check workflow:

```python
ge_data_context_root_dir_with_checkpoint_name_pass = GreatExpectationsOperator(
    task_id="ge_data_context_root_dir_with_checkpoint_name_pass",
    data_context_root_dir=ge_root_dir,
    checkpoint_name="taxi.pass.chk",
)

ge_data_context_config_with_checkpoint_config_pass = GreatExpectationsOperator(
    task_id="ge_data_context_config_with_checkpoint_config_pass",
    data_context_config=example_data_context_config,
    checkpoint_config=example_checkpoint_config,
)

ge_data_context_root_dir_with_checkpoint_name_pass >> ge_data_context_config_with_checkpoint_config_pass
```

> ### Operator Parameters

Reviewer 2: Would it be possible to add an example in this section? Even just a code snippet of what the file structure looks like with the checkpoints, etc. could be helpful. I think you have all the right info here, but as someone not as familiar with GE I think having a visual of some sort would help clarify things.

Author: I tried to link out to examples as much as possible to not clutter the guide because those configs can get pretty long and messy.

For a full list of parameters, see the `GreatExpectationsOperator` [documentation](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator).

By default, a Great Expectations task will run validation and raise an `AirflowException` if any of the tests fail. To override this behavior and continue running even if tests fail, set the `fail_task_on_validation_failure` flag to `false`.

For more information about possible parameters and examples, see the [README in the provider repository](https://github.com/great-expectations/airflow-provider-great-expectations) and the [example DAG in the provider package](https://registry.astronomer.io/dags/example-great-expectations-dag).

### Connections and Backends

> The `GreatExpectationsOperator` can run a checkpoint on a dataset stored in any backend compatible with Great Expectations. All that’s needed to get the Operator to point at an external dataset is to set up an [Airflow Connection](https://www.astronomer.io/guides/connections) to the `datasource`, and add the connection to your Great Expectations project, e.g. [using the CLI to add a Postgres backend](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/database/postgres). Then, if using a `DataContextConfig` or `CheckpointConfig`, ensure that the `"datasources"` field refers to your backend connection name.

Reviewer 2: Not sure if this is feasible, but could we add a list (even a partial one) or what backends are compatible? Like e.g. S3, Snowflake, whatever....
